File: bindings/java/src/test/java/org/wickra/screener/StreamingTest.java
package org.wickra.screener;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

// Streaming equals batch, driven through the JSON command boundary.
//
// screener-core proves this in Rust
// (crates/screener-core/tests/streaming_eq_batch.rs), but that says nothing about
// the boundary each language actually crosses. A binding reaches the core through
// command(), and the golden corpus only ever sends {"cmd":"scan"} -- so feed and
// evaluate were exercised in no language at all. That is how the C ABI shipped a
// command that ran twice under the two-call idiom this binding uses: scan is a
// pure function of its payload, so it looked correct while every feed applied its
// candle twice.
//
// Batch: one scan over the whole dataset.
// Streaming: feed each candle of each symbol in turn, then evaluate.
// Both return the core's compact JSON verbatim, so string equality is the check.
//
// Only the candle-only specs are used. A feeds_* spec needs side feeds the
// streaming envelope carries per bar, and derived_breadth needs the market panel,
// which a streaming screener cannot derive: it sees one symbol's bar at a time and
// cannot know which other symbols print at that timestamp. That is a documented
// limitation (docs/CROSS_SECTION.md), not something to hide here.
class StreamingTest {
    private static final String[] SPECS = {
        "momentum", "mean_reversion", "cross_section_rank", "breadth", "crossover", "compound",
    };

    private static Path findGolden() {
        Path dir = Path.of("").toAbsolutePath();
        for (int i = 0; i < 8 && dir != null; i++) {
            Path g = dir.resolve("golden");
            if (Files.isDirectory(g.resolve("specs"))) {
                return g;
            }
            dir = dir.getParent();
        }
        return null;
    }

    // The binding carries no JSON library -- it is a thin FFI wrapper, and adding
    // a dependency to a test is a dependency all the same. The fixture is
    // machine-generated and its shape is fixed ({"symbol":[{candle}, ...], ...}),
    // so a depth counter that respects strings and escapes is enough. It is
    // guarded by the assertions in the tests: a splitter that returned nothing
    // would fail there rather than pass quietly.
    private static Map<String, List<String>> splitDataset(String json) {
        Map<String, List<String>> out = new LinkedHashMap<>();
        int i = json.indexOf('{');
        while (i >= 0 && i < json.length()) {
            int keyStart = json.indexOf('"', i);
            if (keyStart < 0) {
                break;
            }
            int keyEnd = json.indexOf('"', keyStart + 1);
            String symbol = json.substring(keyStart + 1, keyEnd);
            int arrayStart = json.indexOf('[', keyEnd);
            if (arrayStart < 0) {
                break;
            }
            int arrayEnd = matchingClose(json, arrayStart, '[', ']');
            out.put(symbol, splitObjects(json.substring(arrayStart + 1, arrayEnd)));
            i = arrayEnd + 1;
            int next = json.indexOf(',', i);
            if (next < 0) {
                break;
            }
            i = next;
        }
        return out;
    }

    private static List<String> splitObjects(String arrayBody) {
        List<String> objects = new ArrayList<>();
        int i = 0;
        while (i < arrayBody.length()) {
            int start = arrayBody.indexOf('{', i);
            if (start < 0) {
                break;
            }
            int end = matchingClose(arrayBody, start, '{', '}');
            objects.add(arrayBody.substring(start, end + 1));
            i = end + 1;
        }
        return objects;
    }

    /** Index of the bracket closing the one at {@code open}, respecting strings. */
    private static int matchingClose(String text, int open, char opener, char closer) {
        int depth = 0;
        boolean inString = false;
        for (int i = open; i < text.length(); i++) {
            char c = text.charAt(i);
            if (inString) {
                if (c == '\\') {
                    i++;
                } else if (c == '"') {
                    inString = false;
                }
                continue;
            }
            if (c == '"') {
                inString = true;
            } else if (c == opener) {
                depth++;
            } else if (c == closer) {
                depth--;
                if (depth == 0) {
                    return i;
                }
            }
        }
        throw new IllegalStateException("unbalanced JSON in the golden fixture");
    }

    private static void feedAll(Screener screener, Map<String, List<String>> data) {
        for (Map.Entry<String, List<String>> entry : data.entrySet()) {
            for (String candle : entry.getValue()) {
                screener.command(
                        "{\"cmd\":\"feed\",\"symbol\":\""
                                + entry.getKey()
                                + "\",\"candle\":"
                                + candle
                                + "}");
            }
        }
    }

    @Test
    void streamingEqualsBatch() throws IOException {
        Path golden = findGolden();
        assumeTrue(golden != null, "golden fixtures not present yet");

        String dataText = Files.readString(golden.resolve("data.json")).strip();
        Map<String, List<String>> data = splitDataset(dataText);
        // A splitter that produced nothing would make every comparison trivially
        // true, so say what it found.
        assertTrue(data.size() >= 2, "expected several symbols, got " + data.size());
        for (Map.Entry<String, List<String>> entry : data.entrySet()) {
            assertTrue(
                    entry.getValue().size() >= 10,
                    entry.getKey() + " yielded only " + entry.getValue().size() + " candles");
        }

        int compared = 0;
        for (String name : SPECS) {
            Path specPath = golden.resolve("specs").resolve(name + ".json");
            assertTrue(Files.exists(specPath), "spec " + name + " is missing from the corpus");
            String spec = Files.readString(specPath);

            String batch;
            try (Screener screener = new Screener(spec)) {
                batch = screener.command("{\"cmd\":\"scan\",\"data\":" + dataText + "}").strip();
            }

            String streamed;
            try (Screener screener = new Screener(spec)) {
                feedAll(screener, data);
                streamed = screener.command("{\"cmd\":\"evaluate\"}").strip();
            }

            assertEquals(batch, streamed, "streaming != batch for spec " + name);
            compared++;
        }
        assertEquals(SPECS.length, compared, "not every spec was compared");
    }

    @Test
    void resetReturnsToThePreFeedState() throws IOException {
        Path golden = findGolden();
        assumeTrue(golden != null, "golden fixtures not present yet");

        Map<String, List<String>> data =
                splitDataset(Files.readString(golden.resolve("data.json")).strip());
        String spec = Files.readString(golden.resolve("specs").resolve("momentum.json"));

        try (Screener screener = new Screener(spec)) {
            String empty = screener.command("{\"cmd\":\"evaluate\"}");

            feedAll(screener, data);
            assertNotEquals(
                    empty,
                    screener.command("{\"cmd\":\"evaluate\"}"),
                    "feeding the whole universe changed nothing");

            screener.command("{\"cmd\":\"reset\"}");
            assertEquals(
                    empty,
                    screener.command("{\"cmd\":\"evaluate\"}"),
                    "reset did not return the screener to its pre-feed state");
        }
    }
}

File: examples/c/streaming_test.c
/* Streaming equals batch, through the C ABI's two-call idiom.
 *
 * wickra-screener-core proves this in Rust, but that says nothing about the boundary a
 * C caller crosses. Every reach behind this ABI asks for the response length
 * first and reads it second, and the golden corpus only ever sends
 * {"cmd":"scan"} -- a pure function of its payload. So the command ran twice on
 * every call and nothing noticed: each "feed" applied its candle a second time.
 *
 * The dataset here is written inline rather than read from golden/, because C
 * carries no JSON parser and splitting the corpus by hand would test the
 * splitter. What matters is that the same bars reach the core two ways.
 *
 * The spec uses Sma(3), which has a lookback: a condition over the raw close
 * reads the same however many times a bar arrived, which is precisely how this
 * class of bug stays invisible.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "wickra_screener.h"

static const char *SPEC =
    "{\"universe\":[\"AAA\",\"BBB\"],\"condition\":{\"type\":\"cmp\","
    "\"left\":{\"kind\":\"indicator\",\"name\":\"Sma\",\"params\":[3]},"
    "\"op\":\"gt\",\"right\":{\"kind\":\"const\",\"value\":0.0}}}";

/* Six bars: three per symbol, a rising ramp so Sma(3) is well defined. */
static const char *DATASET =
    "{\"AAA\":["
    "{\"time\":1,\"open\":10,\"high\":10,\"low\":10,\"close\":10,\"volume\":1},"
    "{\"time\":2,\"open\":20,\"high\":20,\"low\":20,\"close\":20,\"volume\":1},"
    "{\"time\":3,\"open\":30,\"high\":30,\"low\":30,\"close\":30,\"volume\":1}],"
    "\"BBB\":["
    "{\"time\":1,\"open\":40,\"high\":40,\"low\":40,\"close\":40,\"volume\":1},"
    "{\"time\":2,\"open\":50,\"high\":50,\"low\":50,\"close\":50,\"volume\":1},"
    "{\"time\":3,\"open\":60,\"high\":60,\"low\":60,\"close\":60,\"volume\":1}]}";

static const char *FEEDS[] = {
    "{\"cmd\":\"feed\",\"symbol\":\"AAA\",\"candle\":{\"time\":1,\"open\":10,\"high\":10,\"low\":10,\"close\":10,\"volume\":1}}",
    "{\"cmd\":\"feed\",\"symbol\":\"AAA\",\"candle\":{\"time\":2,\"open\":20,\"high\":20,\"low\":20,\"close\":20,\"volume\":1}}",
    "{\"cmd\":\"feed\",\"symbol\":\"AAA\",\"candle\":{\"time\":3,\"open\":30,\"high\":30,\"low\":30,\"close\":30,\"volume\":1}}",
    "{\"cmd\":\"feed\",\"symbol\":\"BBB\",\"candle\":{\"time\":1,\"open\":40,\"high\":40,\"low\":40,\"close\":40,\"volume\":1}}",
    "{\"cmd\":\"feed\",\"symbol\":\"BBB\",\"candle\":{\"time\":2,\"open\":50,\"high\":50,\"low\":50,\"close\":50,\"volume\":1}}",
    "{\"cmd\":\"feed\",\"symbol\":\"BBB\",\"candle\":{\"time\":3,\"open\":60,\"high\":60,\"low\":60,\"close\":60,\"volume\":1}}",
};
static const size_t FEED_COUNT = sizeof(FEEDS) / sizeof(FEEDS[0]);

/* Run one command through the documented two-call idiom. Caller frees. */
static char *run(WickraScreener *screener, const char *cmd) {
    int len = wickra_screener_command(screener, cmd, NULL, 0);
    if (len < 0) {
        fprintf(stderr, "command failed: code %d\n", len);
        return NULL;
    }
    char *buf = (char *)malloc((size_t)len + 1);
    if (!buf) {
        return NULL;
    }
    int written = wickra_screener_command(screener, cmd, buf, (size_t)len + 1);
    if (written != len) {
        fprintf(stderr, "second call returned %d, first said %d\n", written, len);
        free(buf);
        return NULL;
    }
    return buf;
}

int main(void) {
    char command[4096];
    int n = snprintf(command, sizeof(command), "{\"cmd\":\"scan\",\"data\":%s}", DATASET);
    if (n < 0 || (size_t)n >= sizeof(command)) {
        fprintf(stderr, "scan command did not fit\n");
        return 1;
    }

    WickraScreener *batch_screener = wickra_screener_new(SPEC);
    if (!batch_screener) {
        fprintf(stderr, "failed to build the batch screener\n");
        return 1;
    }
    char *batch = run(batch_screener, command);
    wickra_screener_free(batch_screener);
    if (!batch) {
        return 1;
    }

    WickraScreener *stream_screener = wickra_screener_new(SPEC);
    if (!stream_screener) {
        fprintf(stderr, "failed to build the streaming screener\n");
        free(batch);
        return 1;
    }
    for (size_t i = 0; i < FEED_COUNT; i++) {
        char *ack = run(stream_screener, FEEDS[i]);
        if (!ack) {
            wickra_screener_free(stream_screener);
            free(batch);
            return 1;
        }
        free(ack);
    }
    char *streamed = run(stream_screener, "{\"cmd\":\"evaluate\"}");
    wickra_screener_free(stream_screener);
    if (!streamed) {
        free(batch);
        return 1;
    }

    int equal = strcmp(batch, streamed) == 0;
    printf("batch    : %s\n", batch);
    printf("streaming: %s\n", streamed);

    /* Sma(3) over 10, 20, 30 is 20 and over 40, 50, 60 is 50. Fed twice each
     * they would be the means of 20, 30, 30 and 50, 60, 60 instead, so these
     * values are what tells the two apart. */
    int has_expected = strstr(batch, "\"Sma(3)\":20.0") != NULL &&
                       strstr(batch, "\"Sma(3)\":50.0") != NULL;

    free(batch);
    free(streamed);

    if (!equal) {
        fprintf(stderr, "streaming != batch\n");
        return 1;
    }
    if (!has_expected) {
        fprintf(stderr, "expected Sma(3) of 20.0 and 50.0 from three bars each\n");
        return 1;
    }
    printf("streaming equals batch\n");
    return 0;
}
